### PR TITLE
Add StringSplitOptions.TrimEntries

### LINKED ISSUE #31038: API proposal: StringSplitOptions.TrimEntries
How many times have we all written the below code?

```cs
string[] entries = someString.Split(',');

for (int i = 0; i < entries.Length; i++)
{
    string entry = entries[i].Trim();
    /* operate on entry */
}
```

Even just a cursory search of our own Framework reference sources shows that this is a common pattern:

 * https://referencesource.microsoft.com/#mscorlib/system/enum.cs,434
 * https://referencesource.microsoft.com/#mscorlib/system/AppContext/AppContextDefaultValues.cs,122
 * https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Annotations/Annotation.cs,665
 * https://referencesource.microsoft.com/#System/compmod/system/codedom/compiler/CodeDomConfigurationHandler.cs,262
 * https://referencesource.microsoft.com/#System/net/System/Net/_AutoWebProxyScriptHelper.cs,582
 * https://referencesource.microsoft.com/#System/net/System/Net/_HeaderInfoTable.cs,84
 * etc.

I propose adding an enum value `StringSplitOptions.TrimEntries` that would make this pattern a bit easier for folks by eliminating the need for developers to call `Trim`.

```cs
[Flags]
public enum StringSplitOptions
{
    None = 0,
    RemoveEmptyEntries = 1,
    TrimEntries = 2 /* new value */
}
```

The behavior of `string.Split` and related APIs would be as follows:

If _None_ or _RemoveEmpyEntries_ is provided, the behavior of `string.Split` under this proposal is unchanged from its behavior today.

If _TrimEntries_ is provided by itself, the `string.Split` method performs the equivalent of calling `string.Trim()` on each element of the returned array. That is, no element in the returned array will have leading or trailing whitespace. The array may contain zero-length strings for entries.

```cs
Assert.Equal(new[] { "Doe", "John" }, "Doe, John".Split(',', StringSplitOptions.TrimEntries));
Assert.Equal(new[] { "1", "", "2", "3" }, "1, , 2 , 3 ".Split(',', StringSplitOptions.TrimEntries));
Assert.Equal(new[] { "", "" }, ", ".Split(',', StringSplitOptions.TrimEntries));
```

If _TrimEntries_ and _RemoveEmptyEntries_ are both provided, the entries are trimmed first, _and then_ zero-length entries are removed from the returned array.

```cs
Assert.Equal(new[] { "Doe", "John" }, "Doe, John".Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries));
Assert.Equal(new[] { "1", "2", "3" }, "1, , 2 , 3 ".Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries));
Assert.Equal(new string[0], ", ".Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries));
```

Adding this enum value may be considered a breaking change if we expect that third-party components aside from `string.Split` are _consuming_ the `StringSplitOptions` enum, not simply generating the value. In theory any such third-party components should be checking their inputs and rejecting unknown flags.

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/31038. Marked __no merge__ since still waiting on API review.

This PR is divided into four commits. It's easiest to review each commit individually.

 1. Add the new flag and ref APIs.
 2. Remove `Utf8StringSplitOptions` and redirect all of the `Utf8String` infrastructure to the improved `StringSplitOptions` type.
 3. Add the `string.Split(..., StringSplitOptions.TrimEntries)` core logic and unit tests.
 4. Find some call sites within _libraries_ to change to use the new API.

I've also done an initial sweep through the __winforms__ (see https://github.com/GrabYourPitchforks/winforms/commit/be2079a840d58a3de619a5165f0cc4de007b39e3) and __aspnetcore__ (see https://github.com/GrabYourPitchforks/aspnetcore/commit/bbf393fae8d940122eda8831952338a2539ab707) repos to see how they can use these APIs.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1281]: Is it worth checking for an empty input string now, in case both TrimEntries and RemoveEmptyEntries is set?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/StringSplitOptions.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/StringSplitOptions.cs:12]: Can you add XML comments for the enum, including the new option?

**Comment by danmosemsft** [comment]
Don't we normally wait for API approval before implementing? Or was this scratching a particular itch 😄

**Comment by stephentoub** [comment]
> Don't we normally wait for API approval before implementing? 

Before putting up a PR, yes :-(

**Comment by GrabYourPitchforks** [comment]
The API was quasi-approved during the Jan 14 review. The only work item from that review was to make sure that callers didn't need `TrimStart` or `TrimEnd` in addition to simply `Trim`. I investigated and found that no callers would benefit from those extra enum entries.

So now that all concerns have been addressed the only thing that's left for API review is approval. Modulo naming if somebody wants a last-minute enum name change. :)

**Comment by danmosemsft** [comment]
I see `dotnet-runtime-perf` ran -- @DrewScoggins would this give us results for perf tests, or just verify the infrastructure didn't break?

Probably should check perf impact on String.Split.

**Comment by GrabYourPitchforks** [comment]
I still need to investigate why some unit tests are failing. The report says that 19 failures occurred, but every test is marked "Pass". So that's fun. :D

I didn't perf test this but I don't anticipate much perf impact. The related PR https://github.com/dotnet/runtime/pull/35733 should slightly improve perf in the `RemoveEmptyEntries` case since it more efficiently resizes the final returned array.

Separately, I was experimenting with a design that removes the intermediate array entirely. That should result in less GC pressure in the `RemoveEmptyEntries` case. But since it introduces some level of complexity and isn't a visible behavioral change I thought it best to reserve that improvement for a different PR.

**Comment by DrewScoggins** [comment]
Those are just correctness perf tests, so not taking real measurements. If we expect some kind of perf impact, either a regression or an improvement, I would recommend that you run performance tests locally. We have this guide, https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md, that should help you get started. If you get stuck on anything feel free to contact either Adam or I. Also once these changes go in we will get lab results and we can take a look at the impact that we are measuring there.

**Comment by danmosemsft** [comment]
Thanks @DrewScoggins I should have realized that (in fact I'm sure you told me that already 😸 )

**Comment by GrabYourPitchforks** [comment]
I'm going to back out the "Remove `Utf8StringSplitOptions` and redirect all of the `Utf8String` infrastructure to the improved `StringSplitOptions` type" part of the PR for now. I didn't fully think through the implications this would have on the downlevel package.

File: src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1281]: Yup! That's actually already done on line 1276.

```cs
        ShortCircuit:
            if (count <= 1 || Length == 0)  // <-- empty input string causes this to be true
            {
                // Per the method's documentation, we'll short-circuit the search for separators.
                // But we still need to post-process the results based on the caller-provided flags.
```

**Comment by GrabYourPitchforks** [comment]
Unit tests were failing due to an off-by-one error in a boundary condition. It's addressed at in the commit https://github.com/dotnet/runtime/pull/35740/commits/5b07dbcf687df27f33d2a76d5dbcc33f17571a02, which is part of this PR. Aside from this and reverting the `Utf8StringSplitOptions` change, the logic is unchanged from the earlier code review.

**Comment by GrabYourPitchforks** [comment]
crossgen2 CI failure is https://github.com/dotnet/runtime/issues/37732

**Pull request merged**
